prov/verbs/src/verbs_info.c

```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
 	return ret;
 }
 
+int vrb_get_port_space(const struct fi_info *info)
+{
+	if (info != NULL && info->addr_format == FI_SOCKADDR_IB)
+		return RDMA_PS_IB;
+	else
+		return RDMA_PS_TCP;
+}
+
+static int vrb_gid_is_null(const union ibv_gid *gid)
+{
+	return !(gid->raw[8]  | gid->raw[9]  | gid->raw[10] | gid->raw[11] |
+	         gid->raw[12] | gid->raw[13] | gid->raw[14] | gid->raw[15]);
+}
+
+void vrb_set_sid(enum rdma_port_space ps,
+        uint16_t port, struct sockaddr_ib *sib)
+{
+        sib->sib_sid = htonll(((uint64_t) ps << 16) + ntohs(port));
+
+        if (ps)
+                sib->sib_sid_mask = htonll(RDMA_IB_IP_PS_MASK);
+        if (port)
+                sib->sib_sid_mask |= htonll(RDMA_IB_IP_PORT_MASK);
+}
+
+static int vrb_get_rdma_addrinfo(struct ibv_context *context,
+			     uint8_t port_num, int tbl,
+			     struct rdma_addrinfo **rai)
+{
+	union ibv_gid gid;
+	uint16_t pkey;
+	int ret;
+
+	ret = ibv_query_gid(context, port_num, tbl, &gid);
+	if (ret)
+		return -FI_EINVAL;
+
+	if (vrb_gid_is_null(&gid))
+		return -FI_EINVAL;
+
+	ret = ibv_query_pkey(context, port_num, tbl, &pkey);
```

Comment:
The gid index and pkey index are independent.

Reply:
Thanks for catching that bug! In case multiple pkeys/gids are set for a port, do we want to create a Libfabric interface for every possible combination of pkeys/gids, or do we want to use the first pair of pkey/gid that is set?